test/rulesets/Base/arraymath.jl

```diff
@@ -15,12 +15,27 @@
         ⋆₂(a, b) = ⋆₂((a, b))  #matrix
         ⋆₂() = ⋆₂(())  # scalar
 
+        ⋆₃(a) = (⋆(n), ⋆(n, 1))
+
         @testset "Scalar-Array $dims" for dims in ((3,), (5,4), (2, 3, 4, 5))
             rrule_test(*, ⋆(dims), ⋆₂(), ⋆₂(dims))
             rrule_test(*, ⋆(dims), ⋆₂(dims), ⋆₂())
         end
 
         @testset "AbstractMatrix-AbstractMatrix" begin
+            # Matrix-Vector product
+            @testset "n=$n, m=$m" for n in (2, 3), m in (4, 5)
+                @testset "Array" begin
+                    rrule_test(*, ⋆(n), ⋆₂(n, m), ⋆₂(m))
+                end
+            end
+
+            @testset "n=$n, m=$m" for n in (2, 3), m in (4, 5)
+                @testset "Array" begin
+                    rrule_test(*, ⋆(n, m), ⋆₃(n), ⋆₂(1, m))
```

Comment:
If we have it implemented correctly then isn't this the correct test?
```suggestion
                    rrule_test(*, ⋆(n, m), ⋆₂(n),  1 ⋆₂ m)
```

Reply:
If we pass in the values that you suggest then we'll get an error like:
```jl
Error During Test at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/check_result.jl:18
  Test threw exception
  Expression: isapprox(add!!(acc_mutated, val), acc + val; kwargs...)
  MethodError: no method matching mul!(::Array{Float64,1}, ::Array{Float64,2}, ::LinearAlgebra.Adjoint{Float64,Array{Float64,2}}, ::Bool, ::Bool)
  Closest candidates are:
    mul!(::AbstractArray, !Matched::Number, ::AbstractArray, ::Number, ::Number) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/generic.jl:124
    mul!(::AbstractArray, ::AbstractArray, !Matched::Number, ::Number, ::Number) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/generic.jl:126
    mul!(::Union{DenseArray{T,1}, Base.ReinterpretArray{T,1,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, Base.ReshapedArray{T,1,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T,1,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, Base.ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}, ::Union{DenseArray{T,1}, DenseArray{T,2}, Base.ReinterpretArray{T,2,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, Base.ReinterpretArray{T,1,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, Base.ReshapedArray{T,1,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, Base.ReshapedArray{T,2,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T,1,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, Base.ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}, SubArray{T,2,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, Base.ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}, !Matched::Union{DenseArray{T,1}, Base.ReinterpretArray{T,1,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, Base.ReshapedArray{T,1,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray}, SubArray{T,1,A,I,L} where L where I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, Base.ReshapedArray{T,N,A,MI} where MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N} where N} where A<:Union{Base.ReinterpretArray{T,N,S,A} where S where A<:Union{SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, SubArray{T,N,A,I,true} where I<:Union{Tuple{Vararg{Real,N} where N}, Tuple{AbstractUnitRange,Vararg{Any,N} where N}} where A<:DenseArray where N where T, DenseArray} where N where T, DenseArray}}, ::Number, ::Number) where T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64} at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/LinearAlgebra/src/matmul.jl:66
    ...
  Stacktrace:
   [1] (::ChainRules.var"#1351#1356"{Array{Float64,2},Array{Float64,2}})(::Array{Float64,1}) at /Users/gaurav/.julia/dev/ChainRules/src/rulesets/Base/arraymath.jl:32
   [2] add!!(::Array{Float64,1}, ::InplaceableThunk{Thunk{ChainRules.var"#1350#1355"{Array{Float64,2},Array{Float64,2}}},ChainRules.var"#1351#1356"{Array{Float64,2},Array{Float64,2}}}) at /Users/gaurav/.julia/dev/ChainRulesCore/src/accumulation.jl:18
   [3] _check_add!!_behavour(::Array{Float64,1}, ::InplaceableThunk{Thunk{ChainRules.var"#1350#1355"{Array{Float64,2},Array{Float64,2}}},ChainRules.var"#1351#1356"{Array{Float64,2},Array{Float64,2}}}; kwargs::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol,Symbol},NamedTuple{(:rtol, :atol),Tuple{Float64,Float64}}}) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/check_result.jl:18
   [4] rrule_test(::Function, ::Array{Float64,2}, ::Tuple{Array{Float64,1},Array{Float64,1}}, ::Vararg{Tuple{Any,Any},N} where N; rtol::Float64, atol::Float64, fdm::FiniteDifferences.FiniteDifferenceMethod{Array{Int64,1},Array{Float64,1},FiniteDifferences.var"#27#28"{FiniteDifferences.FiniteDifferenceMethod{Array{Int64,1},Array{Float64,1},FiniteDifferences.var"#default_bound_estimator#19"{Int64}}}}, fkwargs::NamedTuple{(),Tuple{}}, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/testers.jl:254
   [5] rrule_test(::Function, ::Array{Float64,2}, ::Tuple{Array{Float64,1},Array{Float64,1}}, ::Vararg{Tuple{Any,Any},N} where N) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/testers.jl:225
   [6] top-level scope at In[213]:1
   [7] eval at ./boot.jl:331 [inlined]
   [8] softscope_include_string(::Module, ::String, ::String) at /Users/gaurav/.julia/packages/SoftGlobalScope/u4UzH/src/SoftGlobalScope.jl:217
   [9] execute_code(::String, ::String) at /Users/gaurav/.julia/packages/IJulia/a1SNk/src/execute_request.jl:22
   [10] execute_request(::ZMQ.Socket, ::IJulia.Msg) at /Users/gaurav/.julia/packages/IJulia/a1SNk/src/execute_request.jl:86
   [11] #invokelatest#1 at ./essentials.jl:712 [inlined]
   [12] invokelatest at ./essentials.jl:711 [inlined]

   [13] eventloop(::ZMQ.Socket) at /Users/gaurav/.julia/packages/IJulia/a1SNk/src/eventloop.jl:8
   [14] (::IJulia.var"#15#18")() at ./task.jl:358
  

There was an error during testing

Stacktrace:
 [1] record(::Test.FallbackTestSet, ::Union{Test.Error, Test.Fail}) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:737
 [2] do_test(::Test.ExecutionResult, ::Any) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:520
 [3] _check_add!!_behavour(::Array{Float64,1}, ::InplaceableThunk{Thunk{ChainRules.var"#1350#1355"{Array{Float64,2},Array{Float64,2}}},ChainRules.var"#1351#1356"{Array{Float64,2},Array{Float64,2}}}; kwargs::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol,Symbol},NamedTuple{(:rtol, :atol),Tuple{Float64,Float64}}}) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/check_result.jl:18
 [4] rrule_test(::Function, ::Array{Float64,2}, ::Tuple{Array{Float64,1},Array{Float64,1}}, ::Vararg{Tuple{Any,Any},N} where N; rtol::Float64, atol::Float64, fdm::FiniteDifferences.FiniteDifferenceMethod{Array{Int64,1},Array{Float64,1},FiniteDifferences.var"#27#28"{FiniteDifferences.FiniteDifferenceMethod{Array{Int64,1},Array{Float64,1},FiniteDifferences.var"#default_bound_estimator#19"{Int64}}}}, fkwargs::NamedTuple{(),Tuple{}}, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/testers.jl:254
 [5] rrule_test(::Function, ::Array{Float64,2}, ::Tuple{Array{Float64,1},Array{Float64,1}}, ::Vararg{Tuple{Any,Any},N} where N) at /Users/gaurav/.julia/dev/ChainRulesTestUtils/src/testers.jl:225
 [6] top-level scope at In[213]:1

```

That's the reason to use `⋆₃(n)` instead of `⋆₂(n)`. I'm not sure if I've a mathematical explanation to that, but one thing is that the output `ȳ * xs[2]'` is a matrix not a vector.
So, when the values go into https://github.com/JuliaDiff/ChainRules.jl/blob/master/src/rulesets/Base/arraymath.jl#L32

We need that `X̄` to have dimensions `(2, 1)` not `(2,)`.

Reply:
Then that is the tests telling us the code is wrong.
Differentials always have same shape as primals.

I haven't looked at this closely enough to know why the code is wrong.
We might need another dispatch with `vec` in the code somewhere?